### PR TITLE
Add Neko Wallet adapter

### PR DESCRIPTION
**Pull request by docongminh**
Hi author, @jordansexton 
I added our new `Neko Wallet adapter`. Please review, feedback and merge this if possible for us. 
Thank so much for your supporting us !

### CODE REVIEW

**Comment by docongminh** [comment]
@jordansexton  Hi, When my new-adapter pull request able to merge ?  And Can i need to re-check or prepare done condition to merge it ? 
Thanks !

**Comment by jordansexton** [comment]
Hey, I want to address the lack of activity on your PR. Thank you for your patience!

Here's a followup on the state of wallet-adapter, the issues that individual adapters have in getting integrated with applications, and how we intend to improve things: https://github.com/solana-labs/wallet-adapter/pull/384#issuecomment-1159115764

Let's discuss in that thread to keep it in one place.

**Review by jordansexton — COMMENTED** [review]
While the [medium-term goal](https://github.com/solana-labs/wallet-adapter/pull/384#issuecomment-1159115764) is to move away from requiring adapters for every wallet, I want to respect the time you took to create this adapter, and review it now.

Once a new standard is introduced, let's work together to migrate away from dependency on using an adapter.

File: packages/wallets/neko/src/adapter.ts [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+            // Neko doesn't handle partial signers, so if they are provided, don't use `signAndSendTransaction`
+            if (wallet && 'signAndSendTransaction' in wallet && !options?.signers) {
+                // HACK: Neko's `signAndSendTransaction` should always set these, but doesn't yet
+                transaction.feePayer = transaction.feePayer || this.publicKey || undefined;
+                transaction.recentBlockhash =
+                    transaction.recentBlockhash || (await connection.getRecentBlockhash('finalized')).blockhash;
+
+                const { signature } = await wallet.signAndSendTransaction(transaction, options);
+                return signature;
+            }
```
[the remark this answers is not in the document]
Comment (docongminh) [reply packages/wallets/neko/src/adapter.ts]: Thank for your comment. I fixed it.

**Comment by docongminh** [comment]
@jordansexton Hi, Thank for your support our adapter. I fixed all your comments above and commited new update. Please check our new change. Thank you !

File: packages/wallets/neko/package.json [hunk 2]
```diff
@@ -0,0 +1,33 @@
+{
+    "name": "@solana/wallet-adapter-nekowallet",
```
[the remark this answers is not in the document]
Comment (docongminh) [reply packages/wallets/neko/package.json]: I fixed this problem and changed name: `@solana/wallet-adapter-neko`

File: packages/wallets/neko/package.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    "version": "0.0.1",
```
[the remark this answers is not in the document]
Comment (docongminh) [reply packages/wallets/neko/package.json]: I fixed it and changed version: `0.1.0`

File: packages/wallets/neko/package.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    },
+    "scripts": {
+        "clean": "shx rm -rf lib/*",
+        "build": "yarn clean && tsc -p tsconfig.json && tsc-esm -p tsconfig.json && tsc -p tsconfig.cjs.json",
+        "postbuild": "echo '{\"type\":\"commonjs\"}' | npx json > lib/cjs/package.json && echo '{\"type\":\"module\"} ' | npx json > lib/esm/package.json"
+    },
+    "dependencies": {
+        "@solana/wallet-adapter-base": "^0.9.4",
+        "@solana/web3.js": "^1.36.0"
+    }
```
[the remark this answers is not in the document]
Comment (docongminh) [reply packages/wallets/neko/package.json]: I updated peer/dev deps structure like your comment.

File: packages/wallets/neko/src/adapter.ts [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                            reject(reason);
+                        });
+                    });
+                } catch (error: any) {
+                    if (error instanceof WalletError) throw error;
+                    throw new WalletConnectionError(error?.message, error);
+                } finally {
+                    wallet._handleDisconnect = handleDisconnect;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (docongminh) [reply packages/wallets/neko/src/adapter.ts]: I fixed it with handle connect wallet

File: packages/wallets/neko/src/adapter.ts [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        options?: SendOptions
+    ): Promise<{ signature: TransactionSignature }>;
+    signMessage(message: Uint8Array): Promise<{ signature: Uint8Array }>;
+    connect(): Promise<void>;
+    disconnect(): Promise<void>;
+    _handleDisconnect(...args: unknown[]): unknown;
+}
+
+interface NekoWalletWindow extends Window {
+    solana?: NekoWallet;
```
[the remark this answers is not in the document]
Comment (docongminh) [reply packages/wallets/neko/src/adapter.ts]: I changed to `window.neko`

File: packages/wallets/neko/README.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# `@solana/wallet-adapter-nekowallet`
```
Comment (jordansexton) [inline packages/wallets/neko/README.md:1]:
```suggestion
# `@solana/wallet-adapter-neko`
```

File: packages/wallets/neko/src/adapter.ts [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+declare const window: NekoWalletWindow;
+
+export interface NekoWalletAdapterConfig {}
+
+export const NekoWalletName = 'Neko (Mobile)' as WalletName<'Neko (Mobile)'>;
+
+export class NekoWalletAdapter extends BaseMessageSignerWalletAdapter {
+    name = NekoWalletName;
+    url = 'https://nekowallet.com';
+    icon = 'https://cdn.discordapp.com/attachments/953573404395585596/960393815208914964/Group_7996_1.png';
```
Comment (jordansexton) [inline packages/wallets/neko/src/adapter.ts:50]: This image must be embedded with base64, and should be an SVG. Please heck the other adapters for how this is done.

File: packages/wallets/neko/src/adapter.ts [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    async signMessage(message: Uint8Array): Promise<Uint8Array> {
+        try {
+            const wallet = this._wallet;
+            if (!wallet) throw new WalletNotConnectedError();
+
+            try {
+                const { signature } = await wallet.signMessage(message);
+                return signature;
+            } catch (error: any) {
+                throw new WalletSignTransactionError(error?.message, error);
```
Comment (jordansexton) [inline packages/wallets/neko/src/adapter.ts:195]: This should probably be `WalletSignMessageError` with the corresponding import. Since I'm guessing this code was copied, it probably means there's a problem in the Phantom adapter too.

File: yarn.lock [hunk 10]
```diff
@@ -14819,4 +14819,4 @@ yargs@^16.2.0:
 yocto-queue@^0.1.0:
   version "0.1.0"
   resolved "https://registry.yarnpkg.com/yocto-queue/-/yocto-queue-0.1.0.tgz#0294eb3dee05028d31ee1a5fa2c556a6aaf10a1b"
-  integrity sha512-rVksvsnNCdJ/ohGc6xgPwyN8eheCxsiLM8mxuE/t/mOVqJewPuO1miLpTHQiRgTKCLexL4MeAFVagts7HmNZ2Q==
+  integrity sha512-rVksvsnNCdJ/ohGc6xgPwyN8eheCxsiLM8mxuE/t/mOVqJewPuO1miLpTHQiRgTKCLexL4MeAFVagts7HmNZ2Q==
```
Comment (jordansexton) [inline yarn.lock:14822]: Can you revert the change to this file?

**Comment by docongminh** [comment]
@jordansexton  Hi. I fixed all your comment. Please check and review for me. Thanks !

**Comment by jordansexton** [comment]
Thanks! This will be published today or tomorrow.

**Comment by jordansexton** [comment]
Published! https://twitter.com/jordaaash/status/1541927694786600961

**Comment by docongminh** [comment]
@jordansexton  Thank for your support ! But i checked in [@solana/wallet-adapter-wallets](https://github.com/solana-labs/wallet-adapter/blob/master/packages/wallets/wallets/src/index.ts). Its seem like [@solana/wallet-adapter-neko](https://npmjs.com/package/@solana/wallet-adapter-neko) not yet added to. Do you miss add our adapter ?

**Comment by jordansexton** [comment]
Ah, it's being imported but not exported: https://github.com/solana-labs/wallet-adapter/blob/abac4cd6b661202231c5446493c156000468b429/packages/wallets/wallets/package.json#L44

I'll cut a new release.

**Comment by docongminh** [comment]
@jordansexton  yes. Thank you !

**Comment by jordansexton** [comment]
Done in 29e5c57912f489fd18d5e5240f025d21573551d4 and be8b781d1c3c751af934a05f053be08d15a0b7f2
